File: payment_slips.r

# -----------------------------
# Generate workers
# -----------------------------
generate_workers <- function(num_workers = 400) {

  workers <- data.frame(
    id = 1:num_workers,
    name = paste0("Worker_", 1:num_workers),
    gender = sample(c("Male", "Female"), num_workers, replace = TRUE),
    department = sample(
      c("HRM", "ICT", "Finance", "Actuarial", "Marketing"),
      num_workers,
      replace = TRUE
    ),
    salary = sample(5000:35000, num_workers, replace = TRUE),
    stringsAsFactors = FALSE
  )

  return(workers)
}

# -----------------------------
# Generate payment slips
# -----------------------------
generate_payment_slips <- function(workers) {

  # Print header once
  cat(
    sprintf(
      "%-5s %-12s %-8s %-12s %-8s %s\n",
      "ID", "Name", "Gender", "Department", "Salary", "Level"
    )
  )
  cat(paste(rep("-", 65), collapse = ""), "\n")

  for (i in seq_len(nrow(workers))) {

    tryCatch({

      salary <- workers$salary[i]
      gender <- workers$gender[i]
      department <- workers$department[i]
      level <- "Standard"

      if (salary > 10000 && salary < 20000) {
        level <- "A1"
      }

      if (salary > 7500 && salary < 30000 && gender == "Female") {
        level <- "A5-F"
      }

      cat(
        sprintf(
          "%-5d %-12s %-8s %-12s %-8d %s\n",
          workers$id[i],
          workers$name[i],
          gender,
          department,
          salary,
          level
        )
      )

    }, error = function(e) {
      cat("Unexpected error:", e$message, "\n")
    })
  }
}

# -----------------------------
# Main execution
# -----------------------------
workers <- generate_workers()
generate_payment_slips(workers)

File: payment_slips.py

import random

def generate_workers(num_workers=400):
    workers = []
    for i in range(1, num_workers + 1):
        worker = {
            "id": i,
            "name": f"Worker_{i}",
            "gender": random.choice(["Male", "Female"]),
            "department": random.choice(["HRM", "ICT", "Finance", "Actuarial", "Marketing"]),
            "salary": random.randint(5000, 35000)
        }
        workers.append(worker)
    return workers


def generate_payment_slips(workers):
    # Print header once (Department included)
    print(f"{'ID':<5} {'Name':<12}  {'Gender':<8} {'Department':<12} {'Salary':<8} {'Level'}")
    print("-" * 65)

    for worker in workers:
        try:
            salary = worker["salary"]
            gender = worker["gender"]
            department = worker["department"]
            level = "Standard"

            if 10000 < salary < 20000:
                level = "A1"
            if 7500 < salary < 30000 and gender == "Female":
                level = "A5-F"

            print(
                f"{worker['id']:<5} "
                f"{worker['name']:<12} "
                f"{gender:<8} "
                f"{department:<12} "
                f"{salary:<8} "
                f"{level}"
            )

        except KeyError as e:
            print(f"Missing data for worker: {e}")
        except Exception as e:
            print(f"Unexpected error: {e}")


if __name__ == "__main__":
    workers = generate_workers()
    generate_payment_slips(workers)

